src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.cs

```diff
@@ -103,15 +103,39 @@ public static partial class SymbolFinder
             Solution solution,
             CancellationToken cancellationToken)
         {
-            // If it's already in source, then just return it.
+            // If it's already in source, then we might already be done
             if (InSource(symbol))
             {
-                return symbol;
-            }
+                // If our symbol doesn't have a containing assembly, there's nothing better we can do to map this
+                // symbol somewhere else. The common case for this is a merged INamespaceSymbol that spans assemblies.
+                if (symbol.ContainingAssembly == null)
+                {
+                    return symbol;
+                }
 
-            // If it's not in metadata, then we can't map it back to source.
-            if (!symbol.Locations.Any(loc => loc.IsInMetadata))
+                // Just because it's a source symbol doesn't mean we have the final symbol we actually want. In retargeting cases,
+                // the retargeted symbol is from "source" but isn't equal to the actual source definition in the other project. Thus,
+                // we only want to return symbols from source here if it actually came from a project's compilation's assembly. If it isn't
+                // then we have a retargeting scenario and want to take our usual path below as if it was a metadata reference
+                foreach (var sourceProject in solution.Projects)
+                {
+                    Compilation compilation;
+
+                    // If our symbol is actually a "regular" source symbol, then we know the compilation is holding the symbol alive
+                    // and thus TryGetCompilation is sufficient. For another example of this pattern, see Solution.GetProject(IAssemblySymbol)
+                    // which we happen to call below.
```

Comment:
For #1, your chain is correct: the source symbol holds onto the AssemblySymbol, and SourceAssemblySymbol has the compilation. The compilation being held is literally the first line of code in the class. :-)

For #2, I think they're WeakConstantValueSources or something like that. If you look at the call to solution.GetProject(IAssemblySymbol), you can see the exact same loop and assertion being done.